### ISSUE TITLE
The sync protocol could be described better / I do not understand it

### ISSUE THREAD

**Issue by psychon** [i3/i3#4365]
Hi,

yesterday I rambled a bit about the sync protocol and said that I did not find any docs on it: #4360
(Sorry for all that text)

Today, I found https://i3wm.org/docs/testsuite.html#i3_sync
However, I do not understand this image:
https://i3wm.org/docs/i3-sync-working.png

This shows the sync protocol working:
1. The tests send IPC "focus left" to i3
2. The tests send I3_SYNC to X11
3. The X11 server relays the I3_SYNC to i3
4. i3 handles the IPC "focus left"
5. i3 handles the i3_SYNC

However, what prevents the following sequence of events? (5 happening before 4 in the sequence above)
1. The tests send IPC "focus left" to i3
2. The tests send I3_SYNC to X11
3. The X11 server relays the I3_SYNC to i3
4. i3 handles the i3_SYNC
5. i3 handles the IPC "focus left"

i3 handles the I3_SYNC from the X11 server before it gets the IPC "focus left". As far as I can tell, there is nothing that prevents this kind of race.
(Or to try to make it more explicit: There is a race between i3 handles the IPC "focus left" and replying to the I3_SYNC message. Or yet another way: Nothing establishes a happens-before relationship between these events.)

To actually prevent it, a different kind of protocol would be needed: The tests send something to via IPC to i3, i3 sends that to X11 and X11 sends to the tests.

Is this description somewhat clear? Synchronisation problems are kind of a hobby of mine, but I might have some problems making my thoughts clear.

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow https://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by psychon** [comment]
In the code, I also see an IPC `sync` command. I am not quite sure how it works (I have no prior experience with yajl), but it seems like it does what is needed here: It is an IPC command (thus ensuring that all previous IPC commands have already been executed) that sends an `I3_SYNC` message to an X11 window (thus ensuring that the X11 server already handled all previous X11 requests from i3).

Is there a reason why this is not documented at the same place as the other I3_SYNC mechanism? I also did not find anything making use of this IPC command (but I am not good at looking for it).

- Also, does my reasoning about make sense?
- If so, should I try fixing the races in the test suite?
- If so, would it be okay to just use a heavyweight "sync everything" instead of fine grained syncing? I.e. send an I3_SYNC x11 request and an sync IPC request at the same time. That should ensure all kinds of synchronisation, no matter whether the test suite did an IPC command or an X11 request (I guess the later needs the current `sync_with_i3`).

**Comment by Airblader** [comment]
Thanks for taking such a detailed look at i3 :-)

> Is there a reason why this is not documented at the same place as the other I3_SYNC mechanism?

I don't think so – it probably should.

> I also did not find anything making use of this IPC command

It's used by i3bar, it was introduced in [PR #3214](https://github.com/i3/i3/pull/3214).

> If so, would it be okay to just use a heavyweight "sync everything" instead of fine grained syncing?

To me this sounds OK; racy things like this are difficult to spot and debug, and it's not like this loses significant time, so playing it safe for the tests and "over-syncing" is not an issue to me, but CC @stapelberg since he also introduced these things back then.

**Comment by psychon** [comment]
Ah, sorry, I think I found the missing ingredient. My step 1 above is wrong. I claimed it is:

1. The tests send IPC "focus left" to i3

However, the code actually does:

1. The tests send IPC "focus left" to i3 *and waits for the result*

https://github.com/i3/i3/blob/42c3dbe0255d242a1784ed54e2853c330426c1b9/testcases/lib/i3test.pm.in#L640

This ensures that all X11 requests for "focus left" are already in xcb's output buffer (or perhaps even "already in the kernel's buffer" in case `xcb_flush()` was also called).
Thus, anything "X11-y" that happens later is guaranteed to happen later. Thus, when i3 receives the `I3_SYNC` message from the X11 server, the reply has to happen after the X11 stuff from "focus left".

I still stand by "the docs could be improved". I'll try to take a look at this (eventually). But I'll wait for @stapelberg's comments first.

P.S.: Sorry for all the noise that I have caused.

**Comment by psychon** [comment]
> Thanks for taking such a detailed look at i3 

Sorry for taking such a detailed look at i3 :-)

Don't tell anyone (I'll delete this comment after sending it), but yesterday I switched from AwesomeWM to i3 (to be able to eventually switch to `sway` because wayland). Hence my sudden appearance here.

**Comment by Airblader** [comment]
Please don't apologize, I think you taking the time to look at all of this is very generous!

**Comment by stapelberg** [comment]
Thanks very much for taking such a close look! :) 

> I still stand by "the docs could be improved". I'll try to take a look at this (eventually).

Yes, clarifying that awaiting the IPC response is what establishes the crucial happens-before relationship sounds like a good improvement to me!

Any other improvements to the sync protocol docs are also welcome — I probably should have updated them in PR #3214, sorry about that :)

> * If so, would it be okay to just use a heavyweight "sync everything" instead of fine grained syncing? I.e. send an I3_SYNC x11 request and an sync IPC request at the same time. That should ensure all kinds of synchronisation, no matter whether the test suite did an IPC command or an X11 request (I guess the later needs the current `sync_with_i3`).

If you encounter a test that lacks synchronization, it’s definitely okay to add overly-broad synchronization to make it no longer flaky.

In general, though, I think it’d be even better to only synchronize exactly what’s needed. That way, we don’t waste time in our tests, and we double-check our understanding of which parts are involved in a particular test case.

**Comment by psychon** [comment]
> If you encounter a test that lacks synchronization, it’s definitely okay to add overly-broad synchronization to make it no longer flaky.
>
> In general, though, I think it’d be even better to only synchronize exactly what’s needed. 

I think I found tests with too much synchronisation:

`I3_SYNC` first syncs with the X11 server from the POV of the test suite (since it has to send the ClientMessage to i3, all previous X11 requests from the test suite are processed first), then with i3 (since it handles the ClientMessage), then again with the X11 server from the POV of i3 (same as before, just this time with i3's X11 connection).

The example from https://i3wm.org/docs/testsuite.html#i3_sync does not need this. It did `cmd 'focus left'`, thus it is enough to use the 'new' IPC `sync` command. This command makes i3 send a ClientMessage to some X11 window, thus ensuring that all previous X11 requests from i3 are handled first. This would also make it more obvious why the synchronisation is correct (of course previous IPC commands have to be handled before the next one is handled!).

I didn't read all of the test suite (and I hopefully never will), but so far I think only the tests using XTEST to fake keyboard input need the `I3_SYNC` stuff since they need to make sure the X11 server handled the XTEST requests.

TL;DR: The pattern `cmd 'whatever'; sync_with_i3` IMO syncs too much and it would be enough to use an IPC sync command here instead.

(I will not try implementing this. I already caused enough test failures on CI and flaky tests are way worse than tests syncing too much. Plus, this is again a case of "only minimal gains".)
